Map a resource to multiple signals in eviction manager

jingxu97 opened the pull request:
It is possible to have multiple signals that point to the same type of
resource, e.g., both SignalNodeFsAvailable and
SignalAllocatableNodeFsAvailable refer to the same resource NodeFs.
Change the map from map[v1.ResourceName]evictionapi.Signal to
map[v1.ResourceName][]evictionapi.Signal

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #52661

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
```

jingxu97 committed:
Map a resource to multiple signals in eviction manager

It is possible to have multiple signals that point to the same type of
resource, e.g., both SignalNodeFsAvailable and
SignalAllocatableNodeFsAvailable refer to the same resource NodeFs.
Change the map from map[v1.ResourceName]evictionapi.Signal to
map[v1.ResourceName][]evictionapi.Signal

jingxu97 commented:
@dashpole PTAL

dashpole commented:
/release-note-none

dashpole commented:
Code changes look good.  Im not sure having resourceNodeFs map to both SignalAllocatableNodeFsAvailable and SignalNodeFsAvailable makes sense. This implies that all reclaim functions which free SignalAllocatableNodeFsAvailable also free SignalNodeFsAvailable, and vice versa.  I dont think is true for our reclaim functions.  We have two reclaim functions: Image GC, and Container GC.   But [we only count active (non-terminated) container logs, volumes, and rootfs](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/eviction/helpers.go#L739) towards allocatable.  So cleaning up dead containers, or unused images should **not** increase SignalAllocatableNodeFsAvailable, but only SignalNodeFsAvailable if my reasoning is correct.

What we could do instead would be to have the reclaim functions return a `map[signal]resource.Quantity`  to allow it to specify per-signal the amount that was reclaimed.  Other things may work as well.

jingxu97 commented:
I think the storage resource freed by cleaning up dead containers, or unused images should also increase SignalAllocatableNodeFsAvailable, is there a reason it should not be counted? 

cc @vishh Could you please confirm?

jingxu97 commented:
@vishh Could you please review and approve this? Thanks!

vishh reviewed (changes requested):
Can you add unit tests please?

jingxu97 commented:
@dashpole I changed the code based on you comments. I agree that reclaim nodefs resource should not accounted as allocatable resource.
@vishh I added the unit tests for allocatable resources.

dashpole commented:
lgtm!

vishh commented:
/approve

jingxu97 commented:
/retest

jingxu97 commented:
/retest

jingxu97 commented:
/retest

saad-ali commented:
/test pull-kubernetes-node-e2e

saad-ali commented:
/test pull-kubernetes-e2e-gce-gpu
